R/init.R

```diff
@@ -385,6 +396,55 @@ if (interactive() &&
         page_viewer = page_viewer
       )
 
+      #vsc-r-api
+      response_timeout <- 5
+      response_lock_file <- file.path(dir_session, "response.lock")
+      response_file <- file.path(dir_session, "response.log")
+      file.create(response_lock_file, showWarnings = FALSE)
+      file.create(response_file, showWarnings = FALSE)
+      addin_registry <- file.path(dir_session, "addins.json")
+      # This is created in attach()
+
+      get_response_timestamp <- function() {
+          readLines(response_lock_file)
+      }
+      # initialise the reponse timestamp to empty string 
+      response_time_stamp <- ""
+
+      get_response_lock <- function() {
+        lock_time_stamp <- get_response_timestamp()
+        if (isTRUE(lock_time_stamp != response_time_stamp)) {
+          response_time_stamp <<- lock_time_stamp
+          TRUE
+        } else FALSE
+      }
+
+      request_response <- function(command, ...) {
+        request(command, ..., sd__ = dir_session)
```

Comment:
I guess such quirky argument name `sd__` is used to avoid collision of request arguments. However, we already have `time`, `pid`, `wd`, `command` fields in a typical request, and fields like `id`, `file` are all common names to make it ambiguous to distinguish which are rstudioapi arguments. I think a better way to do this is to capture these arguments in `list(...)`, i.e.

```r
request(command, args = list(...), sd = dir_session)
```

Reply:
Ha I was going to raise an issue to make `wd__` `pid__` etc... but this is better.

Reply:
So your fix is pretty nimble and I do like it, but do you think it is strange that on the handler side in session.ts, for some requests you pull out the arguments with `request.thing`, and others it's `request.arg.thing`?

Reply:
For example:

```
                case 'dataview': {
                    showDataView(request.source,
                        request.type, request.title, request.file, request.viewer);
                    break;
                }
```

Reply:
should these be `request.arg.type`?

Reply:
In fact, I was thinking that the rstudioapi emulation functions should be something like sub-commands. The following call reflects my intention more clearly:

```r
request("rstudioapi", action = "document_save_all", args = list(...), sd = dir_session)
```

Existing commands such as `attach` and `dataview` have known defined arguments. It might make sense that `session.ts` should dispatch commands to the `rstudioapi` level, and those sub-commands should be dispatched in `vsc_r_api.ts`.

Reply:
Yeah this makes sense, I was to thinking about doing this from the very start.

Reply:
Okay that's done.